---
title: "2_Clean OTUs"
author: "Sylvain Piry, Nicolas Rode"
date: "`r format(Sys.Date(), '%d-%B-%Y')`"
output: 
  html_document:
    theme: "journal"
    toc: true
    toc_depth: 3
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options: 
  chunk_output_type: console
---


```{r setup, include =FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
devtools::load_all()
```

# Clean 16S
```{r }
## Follow the instructions to install dada2: https://benjjneb.github.io/dada2/

clean_OTU(abundance_table="16S_Galaxy17-[FROGS_BIOM_to_TSV__abundance.tsv].tsv", affiliation_table="16S_Galaxy18-[FROGS_BIOM_to_TSV__multi-affiliations.tsv].tsv", thresh=0.5, name = "2022_16S_JRLAT", removecol="NC-Index")


```
# Clean COI
```{r }
## Import abundance table
abundance_table <- data.table::fread(file=file.path(here::here(), "data", "raw_data", "COI_Galaxy22-[FROGS_BIOM_to_TSV__abundance.tsv].tsv"),sep="\t",header=TRUE)# to modify (name of the abundance file)

## Import corrected affiliation table
corrected_affiliation_table <- data.table::fread(file=file.path(here::here(), "data", "raw_data", "COI_Galaxy23-FROGS_BIOM_to_TSV__monoaffiliations.csv"),sep=";",header=TRUE)# to modify (name of the abundance file)
head(corrected_affiliation_table)

## Find cluster with multiple affiliation
multi <- abundance_table$observation_name[grep("Multi", abundance_table$blast_taxonomy)]

## Correct affiliation
abundance_table$blast_taxonomy[grep("Multi", abundance_table$blast_taxonomy)] <- corrected_affiliation_table$blast_taxonomy[match(multi, corrected_affiliation_table$`#observation_name`)]

find_chimeras(fordada=abundance_table, name = "2022_COI_JRLAT")

```

# Filter data
```{r }
## Import clean data
abundance_table <- data.table::fread(file=file.path(here::here(), "data", "derived_data", "2022_COI_JRLAT_cleaned_abundance.txt"),sep="\t",header=TRUE)


```
# Update metadata
```{r }
## Import metadata table
metadata <- data.table::fread(file=file.path(here::here(), "data", "raw_data", "Sample_ISS.xlsx - Metadata.csv"), sep=",", header=TRUE)
head(metadata)

## Import clean data
abundance_table <- data.table::fread(file=file.path(here::here(), "data", "derived_data", "2022_COI_JRLAT_cleaned_abundance.txt"),sep="\t",header=TRUE)

## Sum by species
abundance_table_clean <- abundance_table %>% dplyr::select(blast_taxonomy, observation_sum:Undetermined) %>% 
  dplyr::group_by(blast_taxonomy) %>% 
  dplyr::summarise(across(everything(), sum))

##Compute proportion only for Drosophila and replicate A
prop_table_clean <- abundance_table_clean %>% dplyr::filter(grepl("Drosophila", blast_taxonomy)) %>% dplyr::select(blast_taxonomy:Undetermined) %>% dplyr::select(blast_taxonomy, ends_with("-A")) %>%  dplyr::mutate_if(is.numeric, ~round(./sum(.), 1)) 

## Transpose the table
prop_table_clean <- prop_table_clean %>%
    tidyr::gather(key = Barcode, value = value, 2:ncol(prop_table_clean)) %>% 
    tidyr::spread(key = names(prop_table_clean)[1],value = 'value')

## Simplify the name
colnames(prop_table_clean) <- gsub("Arthropoda;Insecta;Diptera;Drosophilidae;Drosophilinae;Drosophila;", "", colnames(prop_table_clean))
## Put back the barcode name
prop_table_clean$Barcode <- gsub("-", "_", prop_table_clean$Barcode)
prop_table_clean$Barcode <- gsub("_A", "", prop_table_clean$Barcode)

## Remove the Drosophila yakuba whose reads are at a very low frequency in the samples
prop_table_clean$Drosophila_yakuba
prop_table_clean <- prop_table_clean  %>% dplyr::select(-Drosophila_yakuba)

## Merge the two datasets
metadata2 <-  dplyr::left_join(metadata, prop_table_clean, by="Barcode")
#metadata2$Drosophila_melanogaster <- as.numeric(metadata2$Drosophila_melanogaster)

write.table(metadata2, file=file.path(here::here(), "data", "derived_data", "Metadata_final.txt"), sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)
```