---
title: "Repeatability_16S"
author: "David Gerin-Jean, Chlo√© Perrais"
date: "`r format(Sys.Date(), '%d-%B-%Y')`"
output: 
  html_document:
    theme: "journal"
    toc: true
    toc_depth: 3
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options: 
  chunk_output_type: console
---


```{r setup, include =FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
devtools::load_all()
```


# Simulating count data in R



# ANOVA, LM, GLM and LMM in R

## Compute repeatability using LMM
```{r }
#importe data
data16S <- data.table::fread(file=file.path(here::here(), "data", "derived_data", "2022_16S_JRLAT_cleaned_filtered_abundance_final.tsv"),sep="\t",header=TRUE)

#split to have family names
data16S_family <- tidyr::separate(data=data16S, col=blast_taxonomy, into=c("kingdom","phylum","class","order","family","genus","species"),  sep=";")
```

##Acetobacter
```{r }
#group each family
dataAAB <- subset(data16S_family, family=="Acetobacteraceae")
#remove useless column
dataAAB_final <- dataAAB[,19:306]
#sum of each row for 1 column
dataAAB_sum_final <- data.frame("count"=Matrix::colSums(dataAAB_final))


library(tidyverse)
dataAAB_sum_final <- dataAAB_sum_final%>%
  rownames_to_column(var="Sample")


dataAAB_sum_final$Sample=gsub("A","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("B","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("C","",dataAAB_sum_final$Sample)



#Linear model
m0 <- lme4::lmer(log(count+1)~1+(1|Sample), data = dataAAB_sum_final)
summary(m0)
## Extract variance components
varcomp <- data.frame(lme4::VarCorr(m0))$vcov
## Compute repeatability
repeatability = varcomp[1]/sum(varcomp)
repeatability
```
 
##Lactobacillaceae
```{r }
#group each family
dataAAB <- subset(data16S_family, family=="Lactobacillaceae")
#remove useless column
dataAAB_final <- dataAAB[,19:306]
#sum of each row for 1 column
dataAAB_sum_final <- data.frame("count"=Matrix::colSums(dataAAB_final))


library(tidyverse)
dataAAB_sum_final <- dataAAB_sum_final%>%
  rownames_to_column(var="Sample")


dataAAB_sum_final$Sample=gsub("A","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("B","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("C","",dataAAB_sum_final$Sample)



#Linear model
m0 <- lme4::lmer(log(count+1)~1+(1|Sample), data = dataAAB_sum_final)
summary(m0)
## Extract variance components
varcomp <- data.frame(lme4::VarCorr(m0))$vcov
## Compute repeatability
repeatability = varcomp[1]/sum(varcomp)
repeatability
```
 
 
 ## Compute repeatability using LMM
```{r }


##Enterobacteriaceae
#group each family
dataAAB <- subset(data16S_family, family=="Enterobacteriaceae")
#remove useless column
dataAAB_final <- dataAAB[,19:306]
#sum of each row for 1 column
dataAAB_sum_final <- data.frame("count"=Matrix::colSums(dataAAB_final))


library(tidyverse)
dataAAB_sum_final <- dataAAB_sum_final%>%
  rownames_to_column(var="Sample")


dataAAB_sum_final$Sample=gsub("A","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("B","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("C","",dataAAB_sum_final$Sample)



#Linear model
m0 <- lme4::lmer(log(count+1)~1+(1|Sample), data = dataAAB_sum_final)
summary(m0)
## Extract variance components
varcomp <- data.frame(lme4::VarCorr(m0))$vcov
## Compute repeatability
repeatability = varcomp[1]/sum(varcomp)
repeatability
```

 
 ## Compute repeatability using LMM
```{r }


##Anaplasmataceae
#group each family
dataAAB <- subset(data16S_family, family=="Anaplasmataceae")
#remove useless column
dataAAB_final <- dataAAB[,19:306]
#sum of each row for 1 column
dataAAB_sum_final <- data.frame("count"=Matrix::colSums(dataAAB_final))


library(tidyverse)
dataAAB_sum_final <- dataAAB_sum_final%>%
  rownames_to_column(var="Sample")


dataAAB_sum_final$Sample=gsub("A","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("B","",dataAAB_sum_final$Sample)
dataAAB_sum_final$Sample=gsub("C","",dataAAB_sum_final$Sample)



#Linear model
m0 <- lme4::lmer(log(count+1)~1+(1|Sample), data = dataAAB_sum_final)
summary(m0)
## Extract variance components
varcomp <- data.frame(lme4::VarCorr(m0))$vcov
## Compute repeatability
repeatability = varcomp[1]/sum(varcomp)
repeatability
```

# Sample abundance table
```{r}
sample_Binomial(Nsamples =100, prob=0.5, mean_nindiviuals_per_habitat=5)

```

=======
# Compute variance when sampling abundance table with only two OTUs (binomial sampling)
## Fixed number of individuals/reads in each sample
```{r}
Nsamples=10000
lambda=50
prob=0.5
## Fixed number of individuals per sample
data <- sample_Binomial(Nsamples=Nsamples, prob=prob, mean_nindiviuals_per_habitat=rep(lambda, Nsamples))
## Estimate probability p with a Binomial glm
m0 <- glm(cbind(OTU1, OTU2)~1, data = data, family="binomial")
summary(m0)

## Observed proportion
(Pobs <- mean(data$OTU1/rowSums(data)))
## Estimated proportion based on logit link
(Pexp=1/(1+exp(-coef(m0))))

## Compute logit(Pexp)
log(Pexp/(1-Pexp))
coef(m0)

## Number of success used for simulations
lambda*prob

## Compute average observed number of success
mean(data$OTU1)

## Compute average expected number of success
mean(rowSums(data))*Pexp

## Compute observed variance
sum((data$OTU1-mean(data$OTU1))^2)/Nsamples

## Compute expected variance
## The formula is available in Wikipedia
lambda*0.5*(1-0.5)
```

## Random number of individuals/reads in each sample
```{r}
## For a general introduction to the issue of estimating the frequency of a sequence using pooled-sequencing, see: Rode et al 2018 https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12723
Nsamples=1000
lambda=50
data <- sample_Binomial(Nsamples=Nsamples, prob=0.5, mean_nindiviuals_per_habitat=lambda)
## Remove rows with zeros as they are never observed in an experiment
data <- data[rowSums(data)!=0,]
## Estimate probability p with a Binomial glm
m1 <- glm(cbind(OTU1, OTU2)~1, data = data, family="binomial")
summary(m1)

## Observed proportion
mean(data$OTU1/rowSums(data))
## Estimated proportion based on logit link
1/(1+exp(-coef(m1)))
## Observed variance of the frequency
sum((data$OTU1/rowSums(data)-mean(data$OTU1/rowSums(data)))^2)/Nsamples
## Expected variance
## Assuming that the original number of bacterial cells is very large and that their is no overdispersion in sequencing depth
## See Gautier et al 2018: https://onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1111%2Fmec.12360&file=mec12360-sup-0002-SupplementMaterialS1.pdf
## Eq. 6 with the number of bacterial cells, n_p, very large and the overdispersion in sequencing depth, s_lambda_p, equals 0): 
(0.5*(1-0.5))*((1+lambda)/lambda^2)
## Cl: please note that the observed and expected variances only match when the number of replicates (i.e. the number of habitats) is large 
```

# Compute alpha and beta diversity when OTU proportions are the same across samples (multinomial sampling)
## Sample abundance table
```{r}
nOTU=3
Nsamples=10
lambda=10
## Sample OTU abundance data
otu_mat <- sample_Multinomial(Nsamples=Nsamples, prob=c(0.1, 0.1, 0.8), mean_nindiviuals_per_habitat=lambda, seed=1)
```

## Create data for Phyloseq analysis
```{r}
## Please add the missing columns here with fake names for Division, Class, ..., Genus
## Create table with taxonomy
tax_mat <- data.frame(otu=paste0("OTU", 1:nOTU), Domain=rep("Eukaryota", nOTU), Supergroup=paste0("Supergroup", 1:nOTU), Division=paste0("Division", 1:nOTU), Class=paste0("Class", 1:nOTU), Order=paste0("Order", 1:nOTU), Family=paste0("Family", 1:nOTU), Genus=paste0("Genus", 1:nOTU))

## Create fake metadata
samples_df <- data.frame(sample=paste0("sample_", 1:Nsamples), TypeofHabitat=as.character(sample(c("Type1", "Type2"), Nsamples, replace = TRUE)), Temperature=as.character(sample(c(17, 20), Nsamples, replace = TRUE)))

## Check number of the two habitats/temperatures
table(samples_df$TypeofHabitat)
table(samples_df$Temperature)
```

## Prepare data for phyloseq analysis
```{r}
## Please follow the instructions here to prepare your data (without subsetting samples or taxa): https://vaulot.github.io/tutorials/Phyloseq_tutorial.html#read-the-data-and-create-phyloseq-objects

##Define the row names from the columns
otu_mat <- otu_mat %>%
  tibble::column_to_rownames("otu") 
tax_mat <- tax_mat %>% 
  tibble::column_to_rownames("otu")
samples_df <- samples_df %>% 
  tibble::column_to_rownames("sample") 
## Transform into matrices otu and tax tables
otu_mat <- as.matrix(otu_mat)
tax_mat <- as.matrix(tax_mat)
## Transform to phyloseq objects
OTU = phyloseq::otu_table(otu_mat, taxa_are_rows = TRUE)
TAX = phyloseq::tax_table(tax_mat)
samples = phyloseq::sample_data(samples_df)

carbom <- phyloseq::phyloseq(OTU, TAX, samples)
carbom
```

## Plot abundance data
```{r}
phyloseq::sample_names(carbom)
phyloseq::rank_names(carbom)
phyloseq::sample_variables(carbom)

## Plot without normalization
phyloseq::plot_bar(carbom, fill = "OTU")

phyloseq::plot_bar(carbom, fill = "Supergroup") + 
  geom_bar(aes(color=Supergroup, fill=Supergroup), stat="identity", position="stack")
    
## Normalize number of reads in each sample using median sequencing depth.
#total = median(phyloseq::sample_sums(carbom))
#standf = function(x, t=total) round(t * (x / sum(x)))
#normalizedcarbom = phyloseq::transform_sample_counts(carbom, standf)
# dev.off()  #if plot_bar not working 

```


## Rarefaction curves
```{r}
p <- phyloseq.extended::ggrare(carbom,
            step = 2,
            color = "TypeofHabitat",
            plot = TRUE,
            parallel = TRUE,
            se = FALSE)
## Define a minimum sample size for rarefaction curve
minSampSize<- min(phyloseq::sample_sums(carbom))

## Split plot by type of habitat
p <- p + 
  facet_wrap(~ TypeofHabitat ) + 
  geom_vline(xintercept = minSampSize, color = "gray60")
p
```

## Plot alpha diversity by sample
```{r}
## use the code available here: https://vaulot.github.io/tutorials/Phyloseq_tutorial.html#bar-graphs)

phyloseq::plot_richness(carbom, measures=c("Observed", "Chao1", "Shannon"))

## Compute Shannon index by hand form Sample 1
P_OTU1 <- otu_mat["OTU1", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU2 <- otu_mat["OTU2", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU3 <- otu_mat["OTU3", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU1+P_OTU2+P_OTU3
H_Sample_1=-(P_OTU1*log(P_OTU1)+P_OTU2*log(P_OTU2)+P_OTU3*log(P_OTU3))

## Compute Shannon index by hand form Sample 3
P_OTU1 <- otu_mat["OTU1", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU2 <- otu_mat["OTU2", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU3 <- otu_mat["OTU3", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU1+P_OTU2+P_OTU3
H_Sample_3=-(P_OTU1*log(P_OTU1)+P_OTU2*log(P_OTU2)+P_OTU3*log(P_OTU3))

## Put the same color of samples from the same type of habitat
phyloseq::plot_richness(carbom, measures=c("Observed", "Chao1", "Shannon"), color="TypeofHabitat")
         
```

## Beta diversity
```{r, eval=TRUE}

dist.a <- phyloseq::distance(carbom, method = "bray")
pa <- phyloseq.extended::plot_dist_as_heatmap(dist.a, show.names = TRUE, title = "Heatmap plot of the beta distance (Bray Curtis) 
between each pair of samples") + 
    theme(plot.title = element_text(hjust = 0.5))
pa

```

## Perform ordination
```{r, eval=TRUE}

carbom.ord <- phyloseq::ordinate(carbom, method ="NMDS", distance ="bray")

## Plot ordination as xyplot     
phyloseq::plot_ordination(carbom, carbom.ord, type="taxa", color="Genus", shape= "Genus", 
                  title="OTUs")

phyloseq::plot_ordination(carbom, carbom.ord, type="samples", color="TypeofHabitat", title="Samples") + geom_point(size=1)

phyloseq::plot_ordination(carbom, carbom.ord, type="split", title="biplot", shape="TypeofHabitat", label = "Temperature") + geom_point(size=3)

## Plot ordination as heat map
phyloseq::plot_heatmap(carbom, method = "NMDS", distance = "bray", 
               taxa.label = "Class", taxa.order = "Class", 
               trans=NULL, low="beige", high="red", na.value="beige")             

phyloseq::plot_heatmap(carbom, method = "NMDS", distance = "(A+B-2*J)/(A+B-J)", 
               taxa.label = "Class", taxa.order = "Class", 
               trans=NULL, low="beige", high="red", na.value="beige")      
```

# Compute alpha and beta diversity when OTU proportions are not the same across samples (multinomial sampling)
## Sample abundance table
```{r}
nOTU=3
Nsamples=10
lambda=10
## Sample OTU abundance data
otu_mat1 <- sample_Multinomial(Nsamples=Nsamples, prob=c(0.1, 0.1, 0.8), mean_nindiviuals_per_habitat=lambda, seed=1)
otu_mat2 <- sample_Multinomial(Nsamples=Nsamples, prob=c(1/3, 1/3, 1/3), mean_nindiviuals_per_habitat=lambda, seed=1)
otu_mat <- cbind(otu_mat1, otu_mat2[,-1])
colnames(otu_mat) = c("otu", paste0("sample_", 1:(Nsamples*2)))

```

## Create data for Phyloseq analysis
```{r}

## Create table with taxonomy
tax_mat <- data.frame(otu=paste0("OTU", 1:nOTU), Domain=rep("Eukaryota", nOTU), Supergroup=paste0("Supergroup", 1:nOTU), Division=paste0("Division", 1:nOTU), Class=paste0("Class", 1:nOTU), Order=paste0("Order", 1:nOTU), Family=paste0("Family", 1:nOTU), Genus=paste0("Genus", 1:nOTU))

## Create fake metadata
samples_df <- data.frame(sample=paste0("sample_", 1:(Nsamples*2)), TypeofHabitat=rep(c("Type1", "Type2"), each=Nsamples), Temperature=as.character(sample(c(17, 20), 2*Nsamples, replace = TRUE)))

## Check number of the two habitats/temperatures
table(samples_df$TypeofHabitat)
table(samples_df$Temperature)
```

## Prepare data for phyloseq analysis
```{r}
## Please follow the instructions here to prepare your data (without subsetting samples or taxa): https://vaulot.github.io/tutorials/Phyloseq_tutorial.html#read-the-data-and-create-phyloseq-objects

##Define the row names from the columns
otu_mat <- otu_mat %>%
  tibble::column_to_rownames("otu") 
tax_mat <- tax_mat %>% 
  tibble::column_to_rownames("otu")
samples_df <- samples_df %>% 
  tibble::column_to_rownames("sample") 
## Transform into matrices otu and tax tables
otu_mat <- as.matrix(otu_mat)
tax_mat <- as.matrix(tax_mat)
## Transform to phyloseq objects
OTU = phyloseq::otu_table(otu_mat, taxa_are_rows = TRUE)
TAX = phyloseq::tax_table(tax_mat)
samples = phyloseq::sample_data(samples_df)

carbom <- phyloseq::phyloseq(OTU, TAX, samples)
carbom
```

## Plot abundance data
```{r}
phyloseq::sample_names(carbom)
phyloseq::rank_names(carbom)
phyloseq::sample_variables(carbom)

## Plot without normalization
phyloseq::plot_bar(carbom, fill = "OTU")

phyloseq::plot_bar(carbom, fill = "OTU") + 
  geom_bar(aes(color=OTU, fill=OTU), stat="identity", position="stack")
    
## Normalize number of reads in each sample using median sequencing depth.
#total = median(phyloseq::sample_sums(carbom))
#standf = function(x, t=total) round(t * (x / sum(x)))
#normalizedcarbom = phyloseq::transform_sample_counts(carbom, standf)
# dev.off()  #if plot_bar not working 

```


## Rarefaction curves
```{r}
p <- phyloseq.extended::ggrare(carbom,
            step = 2,
            color = "TypeofHabitat",
            plot = TRUE,
            parallel = TRUE,
            se = FALSE)
## Define a minimum sample size for rarefaction curve
minSampSize<- min(phyloseq::sample_sums(carbom))

## Split plot by type of habitat
p <- p + 
  facet_wrap(~ TypeofHabitat ) + 
  geom_vline(xintercept = minSampSize, color = "gray60")
p
```

## Plot alpha diversity by sample
```{r}
## use the code available here: https://vaulot.github.io/tutorials/Phyloseq_tutorial.html#bar-graphs)

phyloseq::plot_richness(carbom, measures=c("Observed", "Chao1", "Shannon"))

## Put the same color of samples from the same type of habitat
phyloseq::plot_richness(carbom, measures=c("Observed", "Chao1", "Shannon"), color="TypeofHabitat")
  
## Compute Shannon index by hand form Sample 1
P_OTU1 <- otu_mat["OTU1", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU2 <- otu_mat["OTU2", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU3 <- otu_mat["OTU3", "sample_1"]/sum(otu_mat[, "sample_1"])
P_OTU1+P_OTU2+P_OTU3
H_Sample_1=-(P_OTU1*log(P_OTU1)+P_OTU2*log(P_OTU2)+P_OTU3*log(P_OTU3))

## Compute Shannon index by hand form Sample 3
P_OTU1 <- otu_mat["OTU1", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU2 <- otu_mat["OTU2", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU3 <- otu_mat["OTU3", "sample_3"]/sum(otu_mat[, "sample_3"])
P_OTU1+P_OTU2+P_OTU3
H_Sample_3=-(P_OTU1*log(P_OTU1)+P_OTU2*log(P_OTU2)+P_OTU3*log(P_OTU3))

       
```

## Beta diversity
```{r, eval=TRUE}

dist.a <- phyloseq::distance(carbom, method = "bray")

pa <- phyloseq.extended::plot_dist_as_heatmap(dist.a, show.names = TRUE, order=paste0("sample_", 1:(2*Nsamples)), title = "Heatmap plot of the beta distance (Bray Curtis) 
between each pair of samples") + 
    theme(plot.title = element_text(hjust = 0.5))
pa

```

## Perform ordination
```{r, eval=TRUE}

carbom.ord <- phyloseq::ordinate(carbom, method ="NMDS", distance ="bray")

## Plot ordination as xyplot     
phyloseq::plot_ordination(carbom, carbom.ord, type="taxa", color="Genus", shape= "Genus", 
                  title="OTUs")

phyloseq::plot_ordination(carbom, carbom.ord, type="samples", color="TypeofHabitat", title="Samples", label="TypeofHabitat") + geom_point(size=1)

phyloseq::plot_ordination(carbom, carbom.ord, type="split", title="biplot", color="TypeofHabitat", label = "Temperature") + geom_point(size=3)

## Plot ordination as heat map
phyloseq::plot_heatmap(carbom, method = "NMDS", distance = "bray", 
               taxa.label = "Class", taxa.order = "Class", 
               trans=NULL, low="beige", high="red", na.value="beige")             

phyloseq::plot_heatmap(carbom, method = "NMDS", distance = "(A+B-2*J)/(A+B-J)", 
               taxa.label = "Class", taxa.order = "Class", 
               trans=NULL, low="beige", high="red", na.value="beige")      
```